csi/volumeutils.py

```diff
@@ -27,962 +27,1361 @@
+                    command=MKFS_XFS_CMD
+                ))
 
-def mount_and_select_hosting_volume(pv_hosting_volumes, required_size):
-    """Mount each hosting volume to find available space"""
-    for volume in pv_hosting_volumes:
-        hvol = volume['name']
-        mntdir = os.path.join(HOSTVOL_MOUNTDIR, hvol)
-        mount_glusterfs(volume, mntdir)
+            pvol.save_metadata()
 
-        with statfile_lock:
-            # Stat done before `os.path.exists` to prevent ignoring
-            # file not exists even in case of ENOTCONN
-            mntdir_stat = retry_errors(os.statvfs, [mntdir], [ENOTCONN])
-            with SizeAccounting(hvol, mntdir) as acc:
-                acc.update_summary(mntdir_stat.f_bavail * mntdir_stat.f_bsize)
-                pv_stats = acc.get_stats()
-                reserved_size = pv_stats["free_size_bytes"] * RESERVED_SIZE_PERCENTAGE/100
+        return pvol
 
-            logging.debug(logf(
-                "pv stats",
-                hostvol=hvol,
-                total_size_bytes=pv_stats["total_size_bytes"],
-                used_size_bytes=pv_stats["used_size_bytes"],
-                free_size_bytes=pv_stats["free_size_bytes"],
-                number_of_pvs=pv_stats["number_of_pvs"],
-                required_size=required_size,
-                reserved_size=reserved_size
-            ))
+    @classmethod
+    def _create_subdir_pv(cls, req):
+        """Create a Sub directory based PV"""
+        pvol = PV(name=req.name, type=req.type, pool=req.pool, size=req.size)
+        logging.debug(logf("PV path", pv_path=pvol.path))
 
-            if required_size < (pv_stats["free_size_bytes"] - reserved_size):
-                return hvol
+        # Check for mount availability before creating subdir volume
+        check_mount_availability(pvol.pool.mountpoint)
 
-    return None
+        # Create a subdir
+        makedirs(os.path.join(pvol.pool.mountpoint, pvol.path))
+        logging.debug(logf(
+            "Created PV directory",
+            pvdir=pvol.path
+        ))
 
+        # Write info file so that Brick's quotad sidecar
+        # container picks it up (or) for external quota expansion
+        pvol.save_metadata()
 
-def create_block_volume(pvtype, hostvol_mnt, volname, size):
-    """Create virtual block volume"""
-    volhash = get_volname_hash(volname)
-    volpath = get_volume_path(pvtype, volhash, volname)
-    volpath_full = os.path.join(hostvol_mnt, volpath)
-    logging.debug(logf(
-        "Volume hash",
-        volhash=volhash
-    ))
+        pvol.set_quota()
 
-    # Check for mount availability before creating virtblock volume
-    retry_errors(os.statvfs, [hostvol_mnt], [ENOTCONN])
+        return pvol
 
-    # Create a file with required size
-    makedirs(os.path.dirname(volpath_full))
-    logging.debug(logf(
-        f"Created {pvtype} directory",
-        path=os.path.dirname(volpath)
-    ))
+    @classmethod
+    def create(cls, req):
```

Comment:
- not able to pinpoint exactly the issue however there seems to be something wrong here
- when you create a `PV` you are not linking it to a `Pool` however you are just storing a string version of `Pool` and calling methods off of that, like `check_mount_availability(pvol.pool.mountpoint)`
- In above `mountpoint` is an attr of `pool` but I don't see the linking of `PV` obj to `Pool` obj
- Maybe we need to maintain a global dict of `Pool` objects and link them while creating `PV`(s)
- Pls correct me if i'm missing something (if my deduction is correct, there should be multiple pylint issues which I don't see)

Reply:
Look for `mount_and_select_pool` method to see PV and pool binding.

https://github.com/kadalu/kadalu/pull/868/files#diff-850490ad0b2666d1ea49b540576d0e6bb6525baf9e7b87d84fd841f80758ec4eR709-R718

`req.pool` is already populated by the above method. Check the example given in the commit message https://github.com/kadalu/kadalu/pull/868#issue-1315188973

Reply:
ah! didn't reach there yet I guess, will probably cover some of those in next pass.